### PR TITLE
Force quit after the grace period while closing the Logs Agent.

### PR DESCRIPTION
**Pull request by remeh**
### What does this PR do?

This PR adds a small timer to give some last seconds to the stoppers and then finally force close the Logs Agent.

### Motivation

We use different stoppers (serial/parallel) to stop everything in the Logs Agent in order to not lose any line of log.

However, in some circumstances (reproduced only on Windows) with an Agent monitoring its own log file, when no network is available while closing the Agent, the stoppers never successfully complete.
The Agent is stopping and there is no network: we won't be able to send the logs in anyway.

This PR adds a small timer to give some last seconds to the stoppers and then finally force close the Log Agent.

### Additional Notes

To be more precise, the `input/file/scanner.go` stop method never completes because the decoder of the tailer never flushes properly. The routine is stuck listening on the tailer done channel: https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L199

There may be a better solution as the tailer is also selecting on a `forwardContext` signal, but don't know what can trigger it. https://github.com/DataDog/datadog-agent/blob/master/pkg/logs/input/file/tailer.go#L251

### CODE REVIEW

**Review by blemale — DISMISSED** [review]

**Review by ajacquemot — DISMISSED** [review]
LGTM, thanks for working on this, we could probably lower this value down to 5s. Also this solution does not address the root cause but unblocks customers which is okay but did we plan to analyse where the logs-agent gets stuck ?

File: pkg/logs/agent.go [hunk 1]
```diff
@@ -114,6 +114,13 @@ func (a *Agent) Stop() {
...
 		a.destinationsCtx.Stop()
 		// Wait again for the stopper to complete.
-		<-c
+		// In some situation, the stopper unfortunately never succeed to complete,
+		// we've already reached the grace period, give it some more seconds and
+		// then force quit.
+		timeout := time.NewTimer(10 * time.Second)
+		select {
+		case <-c:
+		case <-timeout.C:
```
Comment (olivielpeau) [inline pkg/logs/agent.go]: let's log something when hit the timeout

File: pkg/logs/agent.go [hunk 2]
```diff
@@ -114,6 +118,39 @@ func (a *Agent) Stop() {
...
+				log.Debug(string(stack))
+			}
+		}
 	}
 }
+
+// getStackTrace returns the stack trace of a running Agent.
+// Used as a best effort to get info while force closing the Logs Agent.
+func getAgentStackTrace() ([]byte, error) {
+	pprofURL := fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/goroutine?debug=2",
```
Comment (ogaca-dd) [inline pkg/logs/agent.go]: Can you be affected by the issue solved by https://github.com/DataDog/datadog-agent/pull/4194?

**Review by ogaca-dd — DISMISSED** [review]
LGTM

File: pkg/logs/agent.go (hunk 2 above)
Comment (remeh) [reply pkg/logs/agent.go]: Nice catch, I think you're right. I'll look into that!

**Review by ogaca-dd — DISMISSED** [review]

**Comment by remeh** [comment]
@ajacquemot I've slightly modified the implementation to log a stack trace if we enter in the "force close" branch of code, in order to have more information to retrieve if it's happening in the wild.
Could you give it a quick glance again? Thank you.

File: pkg/logs/agent.go [hunk 3]
```diff
@@ -114,6 +118,44 @@ func (a *Agent) Stop() {
...
+			} else {
+				log.Debug(string(stack))
+			}
+		}
 	}
 }
+
+// getStackTrace returns the stack trace of a running Agent.
+// Used as a best effort to get info while force closing the Logs Agent.
+func getAgentStackTrace() ([]byte, error) {
```
Comment (ajacquemot) [inline pkg/logs/agent.go]: I feel like we could move this helper somewhere else so that any team can reuse it, WDYT ?

Comment (ajacquemot) [inline pkg/logs/agent.go]: nits: I would change the signature to return a `(string, error)`

Comment (ajacquemot) [reply pkg/logs/agent.go]: nits: it's more a go-routines dump, we could rename this method

**Comment by ajacquemot** [comment]
Just reviewed the PR, looks good, I feel like the go routines dump could be move somewhere else so that it can be reused by any package of the project

File: pkg/logs/agent.go (hunk 3 above)
Comment (remeh) [reply pkg/logs/agent.go]: Absolutely, I've moved the method to `pkg/util` shared with all Agents and changed the function signature to `func GetGoRoutinesDump() (string, error)`

Comment (remeh) [reply pkg/logs/agent.go]: Done!

File: pkg/logs/agent.go [hunk 4]
```diff
@@ -114,6 +115,19 @@ func (a *Agent) Stop() {
...
 		// Wait again for the stopper to complete.
-		<-c
+		// In some situation, the stopper unfortunately never succeed to complete,
+		// we've already reached the grace period, give it some more seconds and
+		// then force quit.
+		timeout := time.NewTimer(5 * time.Second)
+		select {
+		case <-c:
+		case <-timeout.C:
+			log.Debug("Force close of the Logs Agent, dumping the stack trace.")
```
Comment (ajacquemot) [inline pkg/logs/agent.go]: let's update the logs as well :-)

**Review by ajacquemot — DISMISSED** [review]
Let's update the comments accordingly to the last commit, otherwise looks great 🚀

File: pkg/logs/agent.go (hunk 4 above)
Comment (remeh) [reply pkg/logs/agent.go]: Whoops! Done, thank you 🙇

File: pkg/logs/agent.go [hunk 5]
```diff
@@ -114,6 +115,19 @@ func (a *Agent) Stop() {
...
+		// In some situation, the stopper unfortunately never succeed to complete,
+		// we've already reached the grace period, give it some more seconds and
+		// then force quit.
+		timeout := time.NewTimer(5 * time.Second)
+		select {
+		case <-c:
+		case <-timeout.C:
+			log.Debug("Force close of the Logs Agent, dumping the Go routines.")
+			if stack, err := util.GetGoRoutinesDump(); err != nil {
+				log.Debugf("can't get the stack trace of the Agent: %s\n", err)
```
Comment (ajacquemot) [inline pkg/logs/agent.go]: here as well

Comment (remeh) [reply pkg/logs/agent.go]: 🙇

**Review by ogaca-dd — APPROVED** [review]

**Review by ajacquemot — APPROVED** [review]

**Pull request merged**